src/core/models.py

```diff
@@ -0,0 +1,336 @@
+        return self.relation_data.get("fqdn", "")
+
+    @property
+    def ip(self) -> str:
+        """The IP for the unit."""
+        return self.relation_data.get("ip", "")
+
+    @property
+    def server_id(self) -> int:
+        """The id of the server derived from the unit name.
+
+        Server IDs are part of the server strings that ZooKeeper uses for
+        intercommunication between quorum members. They should be positive integers.
+
+        We default to (unit id + 1)
+
+        e.g zookeeper/0 --> 1
+        """
+        return self.unit_id + 1
+
+    @property
+    def host(self) -> str:
+        """The hostname for the unit."""
+        host = ""
+        if self.substrate == "vm":
+            for key in ["hostname", "ip", "private-address"]:
+                if host := self.relation_data.get(key, ""):
+                    break
+
+        if self.substrate == "k8s":
+            host = f"{self.component.name.split('/')[0]}-{self.unit_id}.{self.component.name.split('/')[0]}-endpoints"
+
+        return host
+
+    @property
+    def server_string(self) -> str:
+        """The server string for the ZooKeeper server."""
+        return f"server.{self.server_id}={self.host}:{SERVER_PORT}:{ELECTION_PORT}:participant;0.0.0.0:{CLIENT_PORT}"
+
+    # -- TLS --
+
+    @property
+    def quorum(self) -> str:
+        """The quorum encryption currently set on the unit."""
+        return self.relation_data.get("quorum", "")
+
+    @property
+    def unified(self) -> bool:
+        """Flag to check if server is running with portUnification.
+
+        While switching between TLS + non-TLS server-server quorum encryption,
+        it's necessary to unify the ports first so that members can still
+        communicate during the switch.
+        """
+        return bool(self.relation_data.get("unified", ""))
+
+    @property
+    def private_key(self) -> str:
+        """The private-key contents for the unit to use for TLS."""
+        return self.relation_data.get("private-key", "")
+
+    @property
+    def keystore_password(self) -> str:
+        """The Java Keystore password for the unit to use for TLS."""
+        return self.relation_data.get("keystore-password", "")
+
+    @property
+    def truststore_password(self) -> str:
+        """The Java Truststore password for the unit to use for TLS."""
+        return self.relation_data.get("truststore-password", "")
+
+    @property
+    def csr(self) -> str:
+        """The current certificate signing request contents for the unit."""
+        return self.relation_data.get("csr", "")
+
+    @property
+    def certificate(self) -> str:
+        """The certificate contents for the unit to use for TLS."""
+        return self.relation_data.get("certificate", "")
+
+    @property
+    def ca(self) -> str:
+        """The root CA contents for the unit to use for TLS."""
+        return self.relation_data.get("ca", "")
+
+    @property
+    def sans(self) -> dict[str, list[str]]:
```

Comment:
**suggestion** Why not defining an object or typed dict for this?

I actually found this out because I saw the constants below (e.g. `sans_ip`, `sans_dns`) elsewhere in the code, where we could have just used object attributes/properties

Reply:
We could, but it's barely needed, and any additional types we add introduces additional layers for little value. Leaving it as this is fine.